test/template_projects/base.py

```diff
@@ -0,0 +1,34 @@
+import os
+import io
+import jinja2
+from typing import Union, Dict, Any
+
+
+FilesDict = Dict[str, Union[str, jinja2.Template]]
+TemplateContext = Dict[str, Any]
+
+
+class TemplateProject:
+    files: FilesDict
+    template_context: TemplateContext
+
+    def __init__(self):
+        self.files = {}
+        self.template_context = {}
+
+    def generate(self, path: str):
+        for filename, content in self.files.items():
+            file_path = os.path.join(path, filename)
+            os.makedirs(os.path.dirname(file_path), exist_ok=True)
+
+            with io.open(file_path, 'w', encoding='utf8') as f:
+                if isinstance(content, jinja2.Template):
+                    content = content.render(self.template_context)
+
+                f.write(content)
+
+    def copy(self):
+        other = TemplateProject()
```

Comment:
Alright, I see why it's needed. The differences between C++11/C++14/C++17 tests. Not sure I like this approach best, though.

The (minor, mostly theoretical) issue of not copying the derived class still stands, though. See `c.py` remark for a suggestion that could solve it.

Reply:
> Not sure I like this approach best, though

I'm curious - what about copying the objects don't you like?

Reply:
Hmmm, not entirely sure. I guess I got a bit confused and didn't see the value of copying over just creating a new object, in the example. It's not a huge issue, though. Just more a thought about why this C++ test is so different from the rest (since the other tests don't need `copy`), when it comes to this.